HibernateOptimisticLockingFailureException with s3-log-plugin and 3.3.3

**Describe the bug**

We recently upgrade from Rundeck v3.2.8 to v3.3.3 and now receive the following stack trace upon every job execution and ad hoc command except the _first_ one after the `rundeckd` service is restarted.

```
Exception in thread "pool-2-thread-584" org.springframework.orm.hibernate5.HibernateOptimisticLockingFailureException: Object of class [rundeck.LogFileStorageRequest] with identifier [53048]: optimistic locking failed; nested exception is org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [rundeck.LogFileStorageRequest#53048]
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:284)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:305)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:241)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:120)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave(AbstractHibernateGormInstanceApi.groovy:247)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.save(AbstractHibernateGormInstanceApi.groovy:164)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper.save(GormEntity.groovy:153)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper$save$9.call(Unknown Source)
	at rundeck.LogFileStorageRequest.save(LogFileStorageRequest.groovy)
	at rundeck.LogFileStorageRequest.save(LogFileStorageRequest.groovy)
	at org.grails.datastore.gorm.GormEntity$save.call(Unknown Source)
	at rundeck.services.LogFileStorageService$__tt__runStorageRequest_closure95.doCall(LogFileStorageService.groovy:371)
	at rundeck.services.LogFileStorageService$__tt__runStorageRequest_closure95.doCall(LogFileStorageService.groovy)
	at sun.reflect.GeneratedMethodAccessor2323.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:405)
	at groovy.lang.Closure.call(Closure.java:399)
	at groovy.lang.Closure.run(Closure.java:486)
	at sun.reflect.GeneratedMethodAccessor2062.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:37)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at grails.plugins.executor.PersistenceContextRunnableWrapper$_run_closure1.doCall(PersistenceContextRunnableWrapper.groovy:34)
	at grails.plugins.executor.PersistenceContextRunnableWrapper$_run_closure1.doCall(PersistenceContextRunnableWrapper.groovy)
	at sun.reflect.GeneratedMethodAccessor2061.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:37)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at grails.plugins.executor.PersistenceContextWrapper.wrap(PersistenceContextWrapper.groovy:36)
	at sun.reflect.GeneratedMethodAccessor2060.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:190)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:58)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168)
	at grails.plugins.executor.PersistenceContextRunnableWrapper.run(PersistenceContextRunnableWrapper.groovy:34)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [rundeck.LogFileStorageRequest#53048]
	at org.hibernate.persister.entity.AbstractEntityPersister.check(AbstractEntityPersister.java:2604)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3448)
	at org.hibernate.persister.entity.AbstractEntityPersister.updateOrInsert(AbstractEntityPersister.java:3311)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3723)
	at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:201)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:348)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:102)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1360)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1347)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.flushSession(AbstractHibernateGormInstanceApi.groovy:285)
	at sun.reflect.GeneratedMethodAccessor1194.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1217)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1011)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:994)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi$_performSave_closure3.doCall(AbstractHibernateGormInstanceApi.groovy:250)
	at sun.reflect.GeneratedMethodAccessor1193.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:405)
	at org.codehaus.groovy.runtime.ConvertedClosure.invokeCustom(ConvertedClosure.java:50)
	at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:122)
	at com.sun.proxy.$Proxy193.doInHibernate(Unknown Source)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:297)
	... 53 more
```

We have the [rundeck-s3-log-plugin](https://github.com/rundeck-plugins/rundeck-s3-log-plugin) v1.0.11 installed and configured. It had been working prior to the upgrade from Rundeck v3.2.8. As far as we know, nothing has changed with our S3 configuration. We can see that all three execution logs are successfully uploaded to S3 (`XXX.log.execution.xml`,`XXX.log.rdlog`, and `XXX.log.state.json`). However, only 2 of these are written to the local disk (`XXX.log.rdlog`, and `XXX.log.state.json`).

If we disable the s3-log plugin by commenting out this line `rundeck.execution.logs.fileStoragePlugin=org.rundeck.amazon-s3`, then the stacktraces are not generated and all three log files are written to local disk.

Oddly, we have been unable to reproduce this behavior on our local dev instances of Rundeck, which are Ubuntu 18.04 managed by Vagrant. These have a very similar configuration as our EC2-hosted instances of Rundeck, including using the s3-log plugin to send logs to S3. They are, however, using a local instance of MySQL (v5.7.31).   

We seem to encounter this error on ad hoc commands as well as _all_ of our jobs, including this simple one we have been using to test:

```yaml
- defaultTab: nodes
  description: ''
  executionEnabled: true
  id: 62860dbd-2702-439d-b90e-f2bd1cf1af6e
  loglevel: INFO
  name: uptime
  nodeFilterEditable: false
  plugins:
    ExecutionLifecycle: null
  scheduleEnabled: true
  sequence:
    commands:
    - configuration:
        command: uptime
      nodeStep: true
      type: localexec
    keepgoing: false
    strategy: node-first
  uuid: 62860dbd-2702-439d-b90e-f2bd1cf1af6e
```

**My Rundeck detail**

* Rundeck version: 3.3.3
* install type: deb
* OS Name/version: Ubuntu 18.04.5 LTS
* DB Type/version: MySQL (5.7.26 via Amazon RDS)
* Java: openjdk version "1.8.0_265"

**To Reproduce**

Steps to reproduce the behavior:

1. Run any job. Note the execution ID.
2. Verify log output is generated via web UI.
3. Verify log output is uploaded to S3 for this execution.
4. Verify stacktrace is generated that is similar to the one included above.
5. Inspect local job execution log directory and verify the `XXX.log.execution.xml` file is not present.

**Expected behavior**

The job should run, execution logs (all three) should be uploaded to S3 as well written to the local disk. There should be no corresponding stacktrace.

**Desktop (please complete the following information):**

N/A. This is entirely reproducible outside of the Rundeck web interface.

**Additional context**

We have ruled out there being duplicate entries for the same execution id by running the count query mentioned [here](https://github.com/rundeck/rundeck/issues/5081#issuecomment-620127514).

We are a bit stuck on how to diagnose further. Any recommendations for how to enable more detailed logging or otherwise track this down much appreciated. We <3 Rundeck!

I obtain the same output against MySQL 8 + 1.0.11 plugin on Rundeck 3.3.3.

`framework.properties`:

```
# AWSAccessKeyId and AWSSecretKey can be specified in the file
framework.plugin.ExecutionFileStorage.org.rundeck.amazon-s3.AWSAccessKeyId=KEY-ID
framework.plugin.ExecutionFileStorage.org.rundeck.amazon-s3.AWSSecretKey=SECRET-KEY

# Name of the bucket
framework.plugin.ExecutionFileStorage.org.rundeck.amazon-s3.bucket=rdeck

# Path to store the logs
framework.plugin.ExecutionFileStorage.org.rundeck.amazon-s3.path=logs/${job.project}/${job.execid}.log
```

`rundeck-config.properties`:

```
rundeck.execution.logs.fileStoragePlugin=org.rundeck.amazon-s3
```

I used this Job def (scheduled each 5 secs):

```xml
<joblist>
  <job>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <executionEnabled>true</executionEnabled>
    <id>9e549df5-bad3-4ac3-900e-3e23ca3063e3</id>
    <loglevel>INFO</loglevel>
    <multipleExecutions>true</multipleExecutions>
    <name>HelloWorld</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <plugins />
    <schedule>
      <month month='*' />
      <time hour='*' minute='*' seconds='0/5' />
      <weekday day='*' />
      <year year='*' />
    </schedule>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>
        <exec>echo "hi"</exec>
      </command>
    </sequence>
    <uuid>9e549df5-bad3-4ac3-900e-3e23ca3063e3</uuid>
  </job>
</joblist>
```

Output:

```
Exception in thread "pool-2-thread-4" org.springframework.orm.hibernate5.HibernateOptimisticLockingFailureException: Object of class [rundeck.LogFileStorageRequest] with identifier [48]: optimistic locking failed; nested exception is org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [rundeck.LogFileStorageRequest#48]
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:284)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:305)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:241)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:120)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave(AbstractHibernateGormInstanceApi.groovy:247)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.save(AbstractHibernateGormInstanceApi.groovy:164)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper.save(GormEntity.groovy:153)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper$save$3.call(Unknown Source)
	at rundeck.LogFileStorageRequest.save(LogFileStorageRequest.groovy)
	at rundeck.LogFileStorageRequest.save(LogFileStorageRequest.groovy)
	at org.grails.datastore.gorm.GormEntity$save.call(Unknown Source)
	at rundeck.services.LogFileStorageService$__tt__runStorageRequest_closure95.doCall(LogFileStorageService.groovy:371)
	at rundeck.services.LogFileStorageService$__tt__runStorageRequest_closure95.doCall(LogFileStorageService.groovy)
	at sun.reflect.GeneratedMethodAccessor1787.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:405)
	at groovy.lang.Closure.call(Closure.java:399)
	at groovy.lang.Closure.run(Closure.java:486)
	at sun.reflect.GeneratedMethodAccessor1502.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:37)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at grails.plugins.executor.PersistenceContextRunnableWrapper$_run_closure1.doCall(PersistenceContextRunnableWrapper.groovy:34)
	at grails.plugins.executor.PersistenceContextRunnableWrapper$_run_closure1.doCall(PersistenceContextRunnableWrapper.groovy)
	at sun.reflect.GeneratedMethodAccessor1523.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:37)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at grails.plugins.executor.PersistenceContextWrapper.wrap(PersistenceContextWrapper.groovy:36)
	at sun.reflect.GeneratedMethodAccessor1501.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:190)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:58)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168)
	at grails.plugins.executor.PersistenceContextRunnableWrapper.run(PersistenceContextRunnableWrapper.groovy:34)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [rundeck.LogFileStorageRequest#48]
	at org.hibernate.persister.entity.AbstractEntityPersister.check(AbstractEntityPersister.java:2604)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3448)
	at org.hibernate.persister.entity.AbstractEntityPersister.updateOrInsert(AbstractEntityPersister.java:3311)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3723)
	at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:201)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:348)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:102)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1360)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1347)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.flushSession(AbstractHibernateGormInstanceApi.groovy:285)
	at sun.reflect.GeneratedMethodAccessor938.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1217)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1011)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:994)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi$_performSave_closure3.doCall(AbstractHibernateGormInstanceApi.groovy:250)
	at sun.reflect.GeneratedMethodAccessor896.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:405)
	at org.codehaus.groovy.runtime.ConvertedClosure.invokeCustom(ConvertedClosure.java:50)
	at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:122)
	at com.sun.proxy.$Proxy193.doInHibernate(Unknown Source)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:297)
	... 53 more
```

Note 1: In my environment, the error appears after some executions (not immediately).
Note 2: Also happening on Rundeck 3.3.4 (using `org.mariadb.jdbc.Driver` driver).

Thanks for your feedback @christi3k!

--- a/rundeckapp/grails-app/services/rundeck/services/LogFileStorageService.groovy
+++ b/rundeckapp/grails-app/services/rundeck/services/LogFileStorageService.groovy
@@ -288,98 +288,100 @@ class LogFileStorageService
         failedRequests.remove(requestId)
         failures.remove(requestId)
         long retryMax = 30000;
+        LogFileStorageRequest.withNewSession {
 
-        Execution execution = Execution.get(execId)
+            Execution execution = Execution.get(execId)
+            def files = getExecutionFiles(execution, typelist, false)
 
-        def files = getExecutionFiles(execution, typelist, false)
-        try {
-            def (didsucceed, failuremap) = storeLogFiles(typelist, task.storage, task.id, files)
-            success = didsucceed
-            if (!success) {
-                failures.put(requestId, new ArrayList<String>(failuremap.values()))
-            }
-            if (!success && failuremap && failuremap.size() > 1 || !failuremap[filetype]) {
-                def ftype = failuremap.keySet().findAll { it != null && it != 'null' }.join(',')
-
-                LogFileStorageRequest request = retryLoad(requestId, retryMax)
-                if (!request) {
-                    log.error("Storage request [ID#${task.id}]: Error updating: not found for id $requestId")
-                    success = false
-                } else if (request.filetype != ftype || request.completed != success) {
-                    int retryC = 5
-                    boolean saveDone = false
-                    Exception saveError
-                    while (retryC > 0) {
-                        request = LogFileStorageRequest.get(requestId)
-                        request.refresh()
-                        request.filetype = ftype
-                        request.completed = success
-                        try {
-                            request.save(flush: true)
-                            saveDone = true
-                            break
-                        } catch (Exception e) {
-                            saveError = e
-                            log.debug("Error: ${e}", e)
-                        }
-                        retryC--
-                    }
-                    if (!saveDone) {
-                        log.error("Storage request [ID#${task.id}]: Error updating: $saveError", saveError)
-                    }
+            try {
+                def (didsucceed, failuremap) = storeLogFiles(typelist, task.storage, task.id, files)
+                success = didsucceed
+                if (!success) {
+                    failures.put(requestId, new ArrayList<String>(failuremap.values()))
                 }
+                if (!success && failuremap && failuremap.size() > 1 || !failuremap[filetype]) {
+                    def ftype = failuremap.keySet().findAll { it != null && it != 'null' }.join(',')
 
-            }
-        } catch (IOException | ExecutionFileStorageException e) {
-            success = false
-            log.error("Failure: Storage request [ID#${task.id}]: ${e.message}", e)
-        }
-
-        if (!success && count < retry) {
-            log.debug("Storage request [ID#${task.id}] was not successful, retrying in ${delay} seconds...")
-
-            queueLogStorageRequest(task, delay)
-        } else if (!success) {
-            getStorageFailedCounter()?.inc()
-            if(getConfiguredStorageFailureCancel()){
-                log.error("Storage request [ID#${task.id}] FAILED ${retry} attempts, cancelling")
-                //if policy, remove the request from db
-                executorService.execute {
-                    //use executorService to run within hibernate session
                     LogFileStorageRequest request = retryLoad(requestId, retryMax)
                     if (!request) {
-                        log.error("Storage request [ID#${task.id}]: Error deleting: not found for id $requestId")
-                    } else {
-                        request.delete(flush: true)
-                        log.debug("Storage request [ID#${task.id}] cancelled.")
+                        log.error("Storage request [ID#${task.id}]: Error updating: not found for id $requestId")
+                        success = false
+                    } else if (request.filetype != ftype || request.completed != success) {
+                        int retryC = 5
+                        boolean saveDone = false
+                        Exception saveError
+                        while (retryC > 0) {
+                            request = LogFileStorageRequest.get(requestId)
+                            request.refresh()
+                            request.filetype = ftype
+                            request.completed = success
+                            try {
+                                request.save(flush: true)
+                                saveDone = true
+                                break
+                            } catch (Exception e) {
+                                saveError = e
+                                log.debug("Error: ${e}", e)
+                            }
+                            retryC--
+                        }
+                        if (!saveDone) {
+                            log.error("Storage request [ID#${task.id}]: Error updating: $saveError", saveError)
+                        }
                     }
-                }
-                failures.put(requestId, ["Storage request [ID#${task.id}] FAILED ${retry} attempts, cancelling"])
-                failedRequests.add(requestId)
-            }else{
-                log.error("Storage request [ID#${task.id}] FAILED ${retry} attempts, giving up")
 
-                failures.put(requestId, ["Storage request [ID#${task.id}] FAILED ${retry} attempts, giving up"])
-                failedRequests.add(requestId)
+                }
+            } catch (IOException | ExecutionFileStorageException e) {
+                success = false
+                log.error("Failure: Storage request [ID#${task.id}]: ${e.message}", e)
             }
-        } else {
-            failedRequests.remove(requestId)
-            failures.remove(requestId)
-            //use executorService to run within hibernate session
-            executorService.execute {
-                log.debug("executorService saving storage request status...")
-                LogFileStorageRequest request = retryLoad(requestId, retryMax)
-                if (!request) {
-                    log.error("Storage request [ID#${task.id}]: Error saving: not found for id $requestId")
-                } else if (request) {
-                    log.debug("Loaded LogFileStorageRequest ${requestId} [ID#${task.id}] after retry")
-                    request.refresh()
-                    request.completed = success
-                    request.save(flush: true)
 
-                    log.debug("Storage request [ID#${task.id}] complete.")
+            if (!success && count < retry) {
+                log.debug("Storage request [ID#${task.id}] was not successful, retrying in ${delay} seconds...")
+
+                queueLogStorageRequest(task, delay)
+            } else if (!success) {
+                getStorageFailedCounter()?.inc()
+                if (getConfiguredStorageFailureCancel()) {
+                    log.error("Storage request [ID#${task.id}] FAILED ${retry} attempts, cancelling")
+                    //if policy, remove the request from db
+                    executorService.execute {
+                        //use executorService to run within hibernate session
+                        LogFileStorageRequest request = retryLoad(requestId, retryMax)
+                        if (!request) {
+                            log.error("Storage request [ID#${task.id}]: Error deleting: not found for id $requestId")
+                        } else {
+                            request.delete(flush: true)
+                            log.debug("Storage request [ID#${task.id}] cancelled.")
+                        }
+                    }
+                    failures.put(requestId, ["Storage request [ID#${task.id}] FAILED ${retry} attempts, cancelling"])
+                    failedRequests.add(requestId)
+                } else {
+                    log.error("Storage request [ID#${task.id}] FAILED ${retry} attempts, giving up")
+
+                    failures.put(requestId, ["Storage request [ID#${task.id}] FAILED ${retry} attempts, giving up"])
+                    failedRequests.add(requestId)
                 }
-                getStorageSuccessCounter()?.inc()
+            } else {
+                failedRequests.remove(requestId)
+                failures.remove(requestId)
+                //use executorService to run within hibernate session
+                executorService.execute {
+                    log.debug("executorService saving storage request status...")
+                    LogFileStorageRequest request = retryLoad(requestId, retryMax)
+                    if (!request) {
+                        log.error("Storage request [ID#${task.id}]: Error saving: not found for id $requestId")
+                    } else if (request) {
+                        log.debug("Loaded LogFileStorageRequest ${requestId} [ID#${task.id}] after retry")
+                        request.refresh()
+                        request.completed = success
+                        request.save(flush: true)
+
+                        log.debug("Storage request [ID#${task.id}] complete.")
+                    }
+                    getStorageSuccessCounter()?.inc()
+                }
             }
         }
     }

--- a/rundeckapp/src/test/groovy/rundeck/services/LogFileStorageServiceTests.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/services/LogFileStorageServiceTests.groovy
@@ -88,6 +88,11 @@ class LogFileStorageServiceTests extends HibernateSpec implements ServiceUnitTes
         }
     }
 
+    @org.junit.Before
+    void before() {
+        LogFileStorageRequest.metaClass.static.withNewSession = {Closure c -> c.call() }
+    }
+
     void testConfiguredPluginName() {
         assertNull(service.getConfiguredPluginName())
         service.configurationService=mockWith(ConfigurationService){
@@ -488,6 +493,7 @@ class LogFileStorageServiceTests extends HibernateSpec implements ServiceUnitTes
         return service.getLogFileWriterForExecution(e, [:])
 
     }
+
     void testRunStorageRequestMultiSuccessSingleType(){
         grailsApplication.config.clear()
         grailsApplication.config.rundeck.execution.logs.fileStoragePlugin = "test2"

